### PR TITLE
refactor(eth-client): Remove dependency on `web3`

### PR DESCRIPTION
**Pull request by slowli**
## What ❔

- Removes the dependency on `web3` crate by vendoring its relevant part (types and tokenization traits) into `zksync_basic_types`.
- Refactors `zksync_eth_client` to use `jsonrpsee` instead of `web3` + `jsonrpc-core`.

## Why ❔

`web3` and `jsonrpc-core` are both unmaintained or barely maintained. Using two RPC libraries for L1 / L2 clients is clearly suboptimal.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.
- [x] Spellcheck has been run via `zk spellcheck`.
- [x] Linkcheck has been run via `zk linkcheck`.

### CODE REVIEW

**Review by slowli — COMMENTED** [review]
Couple of general points:

- To minimize the change set, I have not actually unified L1 and L2 client crates in any way. I plan to do it in a follow-up PR.
- For the same reason, I haven't unified any vendored types with `zksync_types::api` types (I suspect at least some of those can be unified).

**Comment by montekki** [comment]
Despite the size of the changeset my question would be does this reduce the amount of future work that still needs to be done in this direction? Cleary as i see it the true path forward is the single framework for L1/L2 which currently looks like it should be
* [`ethers-rs`](https://github.com/gakonst/ethers-rs) for L1
* [`zksync2-rust`](https://github.com/zksync-sdk/zksync2-rust) for L2

The second one being just the necessary extensions of the first one. Generally speaking the move to `ethers-rs` or its modern re-implementation `alloy` is still unavoidable so my question is why do this colossal amount of work that looks like a deadend?

**Comment by slowli** [comment]
@montekki I don't see the effort as wasted; it improves the quality of the code in the repo / its dependencies right now, as `ethers-rs` is already obsolete and `alloy` is not yet mature enough. IMO, postponing such improvements may be not the best course of action since we arguably need good quality of code already (with the EN release, publishing to crates.io and all that). And the amount of work isn't colossal; I'd say that this PR is surprisingly small given that >1,000 LoC are vendored `web3` types.

File: core/lib/eth_signer/src/json_rpc_signer.rs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-use jsonrpc_core::types::response::Output;
+use jsonrpc_core::types::response::Output; // FIXME: remove this dependency
```
[the remark this answers is not in the document]
Comment (popzxc) [reply core/lib/eth_signer/src/json_rpc_signer.rs:1]: I wondered the same thing last I touched it. It was used in Lite, but I think it was never used in Era. Probably it's fine to remove it, but please let's do it as a separate PR -- reviewing PRs that do 10 tangentially related things is very tiresome, as you constantly keep losing track of the essence of the change.

**Review by RomanBrodetski — DISMISSED** [review]
I had a look at the code (including block reverter), it all look good!

Indeed, it can be argued whether it makes sense to switch to jsonrpsee now - instead of waiting for e.g. `alloy`, but I don't see it as a blocker to this PR, as IMO it brings the codebase to a better state

**Review by popzxc — APPROVED** [review]

**Pull request merged**